samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -208,7 +210,8 @@ JobCoordinatorListener createJobCoordinatorListener() {
 
       @Override
       public void onJobModelExpired() {
-        if (container != null) {
+        SamzaContainer container = containerRef.get();
+        if (container != null && containerRef.compareAndSet(container, null)) {
```

Comment:
The `else` block assumes if `containerRef.compareAndSet(container, null)` is false, the containerRef is null. Can we guarantee containerRef doesn't hold another `container` reference?

Reply:
When `SamzaContainerRef` is assigned to new value(other than null), it happens only in `onNewJobModel` in StreamProcessor(Other updations sets it to null: `onContainerStop` & `onContainerFailure`). 

`onJobModelExpired` and `onNewJobModel` will not be executed concurrently, since their executions are serialized through debounceThread. 

I think it's safe to assume that it'll be null in else case(if the compareAndSet fails).

Does that makes sense?

Reply:
Yup makes sense. Thanks for the explanation. I think it'd be useful to document this info in comments since it's not immediately obvious. Might be useful to log for the non-null else case if we come across a future bug through regression or new code.